src/reflect/scala/reflect/internal/Scopes.scala

```diff
@@ -407,14 +407,38 @@ trait Scopes extends api.Scopes { self: SymbolTable =>
 
     override def foreach[U](p: Symbol => U): Unit = toList foreach p
 
-    override def filterNot(p: Symbol => Boolean): Scope = (
-      if (toList exists p) newScopeWith(toList filterNot p: _*)
-      else this
-    )
-    override def filter(p: Symbol => Boolean): Scope = (
-      if (toList forall p) this
-      else newScopeWith(toList filter p: _*)
-    )
+    override def filterNot(p: Symbol => Boolean): Scope = {
+      val first = toList
+      var current = first
+      while ((current ne Nil) && !p(current.head)) {
+        current = current.tail
+      }
+      if (current eq Nil) this
+      else {
+        def notMatch(nextMatch: List[Symbol], firstNonMatch: List[Symbol]): List[Symbol] = {
```

Comment:
I would prefer to depend on the optimized `List.filter` in 2.13.x (which structurally shares the unfiltered suffix) rather than try to re-implement that here.

In 2.12.x, we could avoid double iteration with:

```scala
val result = toList
val filtered = result.filterNot(p)
if (sameLength(result, filtered) this else newScopeWith(filtered: _*)
```